### ISSUE TITLE
Headphone detection

### ISSUE THREAD

**Issue by pmlt** [WebAudio/web-audio-api-v2#89]
**Describe the feature**
When an AudioContext destination is stereo, it would be useful to know whether the output device are headphones or just plain speakers. Knowing about headphones can influence the way stereo panning is done in an application. 

This is most useful for games with positional audio. For example, games with 3D environments could turn on binauralization by default when headphones are detected.

**Is there a prototype?**
Can't prototype this.

**Describe the feature in more detail**
It would simply be a matter of exposing limited information about the nature of the output device given by the host OS. Most platforms offer a way to do this. Here's a couple of examples:

* [Windows Core Audio exposes properties](https://docs.microsoft.com/en-us/windows/win32/coreaudio/core-audio-properties) on audio IMMDevices, including FormFactor, channel configuration mask, and whether the speakers are full-range.
* Android's audio device enumerator returns an [AudioDeviceInfo structure](https://developer.android.com/reference/android/media/AudioDeviceInfo) which tells you the type of output (USB headset, HDMI, etc.)

Of course, headphone detection can't be 100% accurate, especially when the output is through 3.5mm stereo jacks or bluetooth. Therefore, the API should *presume* standard speakers unless a device is explicitely declared to be a headset or headphones (usually via USB).

**Comment by guest271314** [comment]
This could be challenging particularly at Chromium where `enumerateDevices()` is not guaranteed to print the correct device selected or being captured dynamically https://github.com/w3c/mediacapture-main/issues/693.

**Comment by guest271314** [comment]
At Linux where PulseAudio is installed `pavucontrol -t 3` lists output devices.

**Comment by guest271314** [comment]
Related https://github.com/w3c/mediacapture-output/issues/93

**Comment by carlsmith** [comment]
I'm implementing software for a DJ controller, and need to be able to send one signal to the speakers and a different signal to the speakers. Is this possible with the WebAudio API?

**Comment by guest271314** [comment]
@carlsmith 

> I'm implementing software for a DJ controller, and need to be able to send one signal to the speakers and a different signal to the speakers. Is this possible with the WebAudio API?

Do you mean send one signal to speakers a different signal to headphones? 

Am not certain if that is currently possible with Web Audio API alone. What actually occurs might depend on OS; user-agent; if the machine has external speakers or only directs audio output to headphones when headphones are plugged in by default.

**Comment by guest271314** [comment]
At *nix something like this can be used

```
pactl list sinks | grep -E "(Active Port:)" | grep -E "(headphones|lineout)"
```

which, for brevity, can output, for example

```
	Active Port: analog-output-lineout
```

or

```
	Active Port: analog-output-headphones

```

**Comment by guest271314** [comment]
Note, in general, a user can adjust, or move streams. I filed https://github.com/w3c/mediacapture-main/issues/693 for the ability to at least have current data about streams from `getUserMedia()` using `enumerateDevices()`. That issue was closed and this issue https://github.com/w3c/mediacapture-main/issues/730 opened, though the two issues are not the same, nonetheless the second issue was closed as well. 

Re

> Not clear how properties such as label would change, or whether browsers would be able to detect it (apart from polling).

the solution is to call the underlying code (e.g., the PulseAudio instance, `pactl` in the above example) each time `enumerateDevices()` is called. However, since Chromium and Chrome refuses to list monitor devices at Linux https://bugs.chromium.org/p/chromium/issues/detail?id=931749 and the specification does not demand that user-agents list accurate devices, it is not only possible but a fact that Chromium lists the incorrect device when the user dynamically moves the capture stream to Chromium or Chromium input.

Re

> This seems like a user agent configuration issue, and it's not clear there's value in standardizing what they should do here. Closing as no action needed.

there appears to be value here in determining the accurate device that is currently outputting Web Audio API. 

These issues are solvable in the field using existing technologies and workarounds. Specification content reflecting the reality in the implementation, user-agent interoperability with the same code, and design of API's at the outset with the capacity to expand beyond the scope of initial concepts, which could very well be limited, or broad yet missing details observed during the process, is a different exercise than achieving a requirement without waiting on any individual or institution involved in the first part of this sentence. In the meantime we can create proof-of-concepts of the features requested, and when they work as intended, just use the code you rolled yourself, using answers to questions that might have already been asked, or that you answer yourself.

We do not need to poll. We just need an API which outputs the correct information in spite of user-agent internal refusal to implement a given feature or standard.

**Comment by carlsmith** [comment]
> Do you mean send one signal to speakers a different signal to headphones?

@guest271314, yep, sorry - I've updated the question.

**Comment by carlsmith** [comment]
> there appears to be value here in determining the accurate device that is currently outputting Web Audio API.

`+1` The essential functionality of any DJ software (for mixing/scratching or for broadcasting/podcasting) depends on being able to route audio between specific input and output devices.

> Am not certain if that is currently possible with Web Audio API alone. What actually occurs might depend on OS; user-agent; if the machine has external speakers or only directs audio output to headphones when headphones are plugged in by default.

Could it be made to work reliably with something like Electron on Linux, assuming I have full control over the system? I will have to abandon my project otherwise.

To be honest, the ability to output one thing to the speakers, and another thing to the headphones seems like a pretty reasonable expectation from a general purpose audio API.

The Mixxx developers would know a lot about the technicalities, at least for the desktop operating systems, but I don't want to ask for their advice, if the API developers are not willing to implement it anyway.

**Comment by bradisbell** [comment]
We can set a sink on an HTMLMediaElement.  The specifics of this API are being modified, but we just need to make sure that same capability exists for an AudioContext.

**Comment by guest271314** [comment]
> We can set a sink on an HTMLMediaElement.

At Chromium or Chrome on Linux that will always only be a microphone, a sink, not actual audio output https://github.com/w3c/mediacapture-output/issues/111. AFAIK there is no standardized API to select any and all devices, e.g., https://github.com/w3c/mediacapture-main/issues/703#issuecomment-653822239; https://github.com/w3c/mediacapture-main/issues/720.

We can select any source output or sink without a specification right now using Native Messaging and, or `WebTransport`; the capability exists, e.g., a (WIP; am not done testing, as the result is not consistently the same as just running `./toggle_source.sh <device>` at the terminal - sometimes have to run the code twice, and still selects the incorrect device, e.g., a USB device that I hot-plugged in for testing, where at the terminal the same shell script toggles correct at each execution)

```
async function toggle_source(source = 'Description: Monitor of Built-in Audio Analog Stereo') {
  const quicTransportServerURL = `quic-transport://localhost:4433/toggle_source`;
  try {
    const transport = new WebTransport(quicTransportServerURL);
    console.log(transport);
    transport.onstatechange = async (e) => {
      console.log(e);
    };
    await transport.ready;
    transport.closed
      .then((reason) => {
        console.log('Connection closed normally.', { reason });
      })
      .catch((e) => {
        console.error(e.message);
        console.trace();
      });
    const sender = await transport.createUnidirectionalStream();
    const writer = sender.writable.getWriter();
    const encoder = new TextEncoder('utf-8');
    let data = encoder.encode(source);
    await writer.write(data);
    const reader = transport.incomingUnidirectionalStreams.getReader();
    console.log({ transport, sender, reader });
    const result = await reader.read();
    if (result.done) {
      console.log(result);
      return;
    }
    let stream = result.value;
    console.log({ stream });
    const { readable } = stream;
    console.log(
      await readable.pipeThrough(new TextDecoderStream()).getReader().read()
    );
    console.log({ reader, transport });
    await reader.cancel();
    transport.close();
    console.log(await reader.closed, await stream.readingAborted);
  } catch (e) {
    console.warn(e.message);
    throw e;
  }
}
```

```  
# ..              
data = subprocess.run(['./toggle_source.sh', input_data], stdout=subprocess.PIPE) 
# ..
```

```
#!/bin/bash
SOURCE_OUTPUT_INDEX=`pactl list source-outputs | grep -E "(Source\sOutput|media\.name)" | grep Record\Stream -B1 | grep -o -E -m1 "[0-9]+"`
SOURCE_INDEX=`pactl list sources | grep -E "(Source\s|Description:\s)" | grep "$1" -B1 | grep -o -E -m1 "[0-9]+"`
pactl move-source-output $SOURCE_OUTPUT_INDEX $SOURCE_INDEX
echo 'ok'
```

and since Web Speech API does not support capture of speech synthesis audio output, we do not use Web Speech API at all and get the WAV directly from `espeak-ng` https://github.com/guest271314/quictransport/blob/main/speech-synthesis/quicTransportEspeakNG.js and use Web Audio API to output (though note, that API is in transition; `QuicTransport` is obsolete). We should be able to select `speech-dispatcher-espeak-ng` playback device using one of these API's, though the reality is even though I and other developers in the field have filed numerous bugs and issues, and demonstrated the viability of the algorithm or process described, the words do not get written out and implementers do what they want. So can developers in the field; authoring specifications is merely the science of words, how that even plays out in code might be different, and in flux; roll your own in any event.

**Comment by guest271314** [comment]
In theory `navigator.mediaDevices.ondevicechange` should be able to handle this https://github.com/w3c/mediacapture-main/issues/730#issuecomment-712513769

> I think we have to be careful not to disturb what I think is the primary use case of devicechange:
>
> - Detecting user insertion (or bluetooth activation) of a new device (meaning: user intends to use this device)

though does not, see above linked issues.

Although specification authors are want to not create overlapping API's or methods we can use the Media Cpature and Stream model with `AudioContext`, something like `AudioContext.ondevicechange` and `AudioContext.ondestinaionchange` if we want an event pattern (PulseAudio is capable of events), or `AudioContext.getDestination()` if we want to get the current accurate destination. That means that Web Audio API must ignore any restrictions imposed by other API or omitted from other API's to get and deliver the correct information to the user, for example, the user dynamically changes the destination to some device that is not specified to be an output device anywhere this API must still deliver the correct data to the user, else this API will be no different than existing API's that deliver incorrect information, and know they are doing such. This is all very possible right now, and we, the developers ourselves, can create this proof-of-concept for each respective platform, and perform the necessary testing, the "deliverable", then the specification authors' only chore here is to write it out.

**Comment by guest271314** [comment]
FWIW At Chromium or Chrome on *nix where `pulsesaudio` is installed it should be possible to modify this code to detect headphones or speakers, and set and output or input devices, create virtual devices, etc., or basically do whatever you want with audio as a `MediaStreamTrack` outside of specifications or user-agent implemented API's https://github.com/guest271314/setUserMediaAudioSource. Cheers.

**Comment by hoch** [comment]
Selecting an audio device can be done with [MediaDevices.enumerateDevices()](https://developer.mozilla.org/en-US/docs/Web/API/MediaDevices/enumerateDevices) and MediaStreamAudioDestinationNode. This issue is not about selecting an output, but detecting a device type of the selected audio device. What I propose is to utilize [MediaDeviceInfo](https://w3c.github.io/mediacapture-main/#ref-for-dom-mediadeviceinfo-11).

```webidl
[Exposed=Window, SecureContext]
interface MediaDeviceInfo {
  readonly attribute DOMString deviceId;
  readonly attribute MediaDeviceKind kind;
  readonly attribute DOMString label;
  readonly attribute DOMString groupId;
  [Default] object toJSON();
};
```

Perhaps we can extend [MediaDeviceKind](https://w3c.github.io/mediacapture-main/#dom-mediadevicekind):

```webidl
enum MediaDeviceKind {
  "audioinput",
  "audiooutput",
  "audiooutput-speakers",
  "audiooutput-headsets",
  "videoinput"
};
```

In terms of the precision of detection, I like @pmlt's approach here:
> Therefore, the API should presume standard speakers unless a device is explicitly declared to be a headset or headphones (usually via USB).

Additionally, the API can return `speakers` or `headsets` only when it's certain - avoiding false positive.

I see 2 problems:
1. Building a precise detection logic can be a real challenge here, but I think we should follow the platform-provided API for the first cut.
2. The relevant spec change will happen outside of this WG/CG.

**Comment by bradisbell** [comment]
@hoch Can you clarify whether an unknown output type would be `audiooutput` or `audiooutput-speakers`?  It's often unknown what the device type is.  As a web developer, I would rather have plain `audiooutput` and not have the device indicated as a speaker unless there was certainty that it was.

I think that's what you're getting at, but it seems to be slightly different from @pmlt 's comment that you quoted.

**Comment by hoch** [comment]
Yes. Building a consensus on the detection logic would be a huge part of this issue.

>  As a web developer, I would rather have plain audiooutput and not have the device indicated as a speaker unless there was certainty that it was.

That's what I mentioned in my comment:

> the API can return speakers or headsets only when it's certain

What's the slight difference you see? I would like to know.

**Comment by guest271314** [comment]
Note, there is a problem with the current `"audiooutput"` described in Media Capture and Streams and `setSinkId()` in the Audio Output API: That is **not** actually audio output https://github.com/guest271314/SpeechSynthesisRecorder/issues/18, rather _playback_, so we cannot rely on that definition for real audio output. Mozilla does not use that term of art for input devices.

**Comment by guest271314** [comment]
`navigator.mediaDevices.ondevicechange` theoretically can handle notification of device changes, however that event does not provide any meaningful information about which device changed via physically or virtual connection to the machine. A user would need to check which device is being captured before and after a device change.

The system can detect if a device is described as headphones or speakers or line-out. For `ondevicechange` to be effective for this use case a `currentDevice` and, or `lastDevice` property could be added to the event so that a user or application can gain useful information from the event.

If a user tries to select `"audiooutput"` on Chrome or Chromium with `getUserMedia()` they will only ever select a microphone device. Try this code
       
        navigator.mediaDevices.getUserMedia({audio: true})
        .then(stream => navigator.mediaDevices.enumerateDevices()
        .then(devices => {
          const audiooutput = devices.find(device => device.kind == "audiooutput");
          stream.getTracks().forEach(track => track.stop());
          if (audiooutput) {
            const constraints = {
              deviceId: {
                exact: audiooutput.deviceId
              }
            };
            return navigator.mediaDevices.getUserMedia({
              audio: constraints
            });
          }
          return navigator.mediaDevices.getUserMedia({
            audio: true
          });
        }));


for yourself and post here is any sound other than microphone is captured.

I guarantee you that at Chrome or Chromium default stable shipment that there is no difference between setting the `exact` constraint to the `deviceId` of the device labelled `"audiooutput"` and the default deviced labelled `"audioinput"`. 

Therefore the question must be asked: What is the technical and effective difference between `"audiooutput"` and `"audioinput"` device descriptions on Chrome or Chromium?

If this question cannot be answered for disambiguation `"audiooutput"` should not be relied on for anything other than a microphone https://bugs.chromium.org/p/chromium/issues/detail?id=1114422#c7

> For gUM, the only field of an audiooutput device you can use is the groupID, which can be used to select a microphone (i.e. audioinput kind) with the same group ID, if available.

which is obviously contrary to and inconsistent with the plain meaning of the term in two words audio output, leading to confusion when a user might take the term `"audiooutput"` literally [This is again recording from microphone, not from audiooutput device #14](https://github.com/guest271314/SpeechSynthesisRecorder/issues/14), e.g., see the comments at this [answer](https://stackoverflow.com/a/45003549) 

> Hi @guest271314, isn't this recording the user's mic - and not the actual synthesized speech? Is that what you intended? – Ronen Rabinovici Dec 1 '17 at 11:26 

> 2 @loretoparisi See updated code which sets media device to record to "audiooutput" plnkr.co/edit/PmpCSJ9GtVCXDhnOqn3D?p=preview – guest271314 Jan 10 '18 at 3:22

> 2 @guest271314, I used the code at plnkr.co/edit/PmpCSJ9GtVCXDhnOqn3D?p=preview but it still recorded from my microphone. – Jeff Baker Aug 15 '18 at 22:54

> This doesn't record speaker output. I tried capturing tab audio using chrome extension but still failed. It seems speechSynthesis is not using HTMLmediaElement for audio hence we shall not be able to capture at tab/browser level. The audiooutput mentioned above returns "default " for both mic and speaker since there is no way to set "kind" field while setting constraints in getUsermedia, it always captures "mic". Let me know in case more details required. – Gaurav Srivastava Mar 4 '19 at 1:13

> Confirming that it records from microphone rather than speech synthesis - at least in Chrome 84. – joe Aug 13 '20 at 11:15

as microphone is not generally construed as output, rather as input, which is the only subject matter Media Capture and Streams deals with, in spite of the fact that it is possible to create virtual devices locally that are not microphones and capture that real audio output, or whatever you want, at `getUserMedia()` by selecting the label you created locally.

So, I would not use the term `"audiooutput"` if you are actually talking about audio output and developers should not take that term to mean such -as to _capture_ in the Media Streams and Capture specification.

For example, I filed this issue [Default device selection and listing inconsistencies #727](https://github.com/w3c/mediacapture-main/issues/727) because the device listing between Firefox and Chrome is inconsistent .

The issue was closed based on this comment https://github.com/w3c/mediacapture-main/issues/727#issuecomment-740050327

> Since it is impossible to define a mechanism to determine default devices that works for all platforms and device kinds, I don't think there is much that the spec can do beyond the current language encouraging to use system defaults.

Then the same individual that posted that comment filed an issue that refutes their own statement as to the task of listing of default and just devices in a specific order being impossible 
[default audio output should be first in the enumerateDevice returned list #756](https://github.com/w3c/mediacapture-main/issues/756)

> The spec already mandates that the default microphone and the default camera (if they exist) should be the first devices listed in enumerateDevices. This should also apply to speakers.

yet that is the exact opposite rationale used to close the former issue. If it is impossible to determine default devices, then how can you set any default device as first in the list?

The current status of that issue is https://github.com/w3c/mediacapture-main/issues/756

> Reopening, we need a WPT test probably

yet the comments at the SO answer provide all the information needed, no matter how many times the supposed `"audiooutput"` device is tested on Chrome and Chromium - again, Firefox and Nightly do not support that label - a _**microphone**_ will be captured - _not_ actual system audio _output_.

We need to do something different here at Web Audio API if we are to be transparent about what is really going on here, to avoid the issue of just creating arbitrary terms of art that in practical coding are contrary to plain meaning of the terms used and that eventually exposes the contradiction.

As to this issue, we should be able to 

1) determine the initial type of device the audio is being output on
2) create an event to notify the user if an when the device changes to non-headphone device

We cannot rely on the existing terms in Audio Output Devices API nor Media Capture and Streams specification for this API due to the above listed contradictions.

**Comment by guest271314** [comment]
Where "the rubber meets the road" re https://github.com/w3c/mediacapture-main/issues/756

> Reopening, we need a WPT test probably

is [Issue 865799 in chromium: A way for enumerateDevices() to show audiooutput devices with labels when user has no cam or mic](https://bugs.chromium.org/p/chromium/issues/detail?id=865799#c25) where on the use case is _listening_ to a Google Meet session without having a microphone attached - yet Chrome or Chromium when a user does not have a microphone device attached to the machine Chrome will throw a device "not found" or "refuse to start device error" - meaning on Chrome `"audiooutput"` is not really "speakers or headphones" at all, and Chrome refuses to capture non-microphone devices. 

Now, if `"audiooutput"` actually meant the plain language audio output user should be able to capture the audio output and listen in on the Google Meet or Google Hangouts session - without a microphone attached - the reality is that is currently impossible on Chrome, and setting the default `"audiooutput"` `MediaDeviceInfo` as first on the list will not change that fact, as Chrome only _captures_ `InputDevice`'s.

So, no many how many tests are run, nor what label is applied at the implementation label, `"audiooutput"` technically still means microphone for all practical and technical purposes.

**Comment by guest271314** [comment]
> the reality is that is currently impossible on Chrome,

Correction: Not impossible, yet not specified and requires the user to define the virtual device or remap devices so that Chrome does not "see" "Monitor" on Linux - again, which Chrome refuses to capture https://github.com/guest271314/captureSystemAudio#pulseaudio-module-remap-source.

**Comment by hoch** [comment]
A slight modification from https://github.com/WebAudio/web-audio-api-v2/issues/89#issuecomment-756234076:

```webidl
[Exposed=Window, SecureContext]
partial interface MediaDeviceInfo {
  readonly attribute MediaDeviceSubkind subkind;
};
```

```webidl
enum MediaDeviceSubkind {
  "unknown",
  "speakers",
  "headsets",
};
```

This ensures backward compatibility and does not bleed out to the existing enum.

I want @pmlt's input on this because if we have data of the platform-provided detection API it would be easier to evaluate the utility of this feature. If "it works only when it is certain" and that "when" is less than 0.1%, then we really have to give a second thought on this.

I was also thinking about being able to detect headphones and use it to activate/deactivate the AEC, but this is not really practical according to the Chromium's internal discussion. (mostly due to the faulty device) So that's another reason to give a second thought.

**Comment by pmlt** [comment]
That would work. If the user selects an output device whose nature/form factor cannot be detected by the system, then the application receives the "unknown" MediaDeviceSubkind. When it does, it can default to whatever it wishes, or even offer the user the choice to select between an audio mix optimized for TV speakers, headphones, or surround (as many games often do in their audio options menu).

**Comment by hoch** [comment]
So you think this change will be beneficial when you will end up with "unknown" for 90% of the time?

**Comment by pmlt** [comment]
It will be beneficial for the 10% of cases where it works :) Whether the benefits are worth the spec and development time is another question altogether. I would be fine with this issue being closed as "Won't Do" now that we've properly studied the use case.

**Comment by hoch** [comment]
I am still researching. So far, the only platform supports something close to this idea is Android. It attempts to detect the impedance of a device plugged in and tell you speakers or headphones. Given that newer mobile phones don't even have a mini phono jack these days, the beneficial that is already small will be declining in the future.

Have you tried to work with MediaDevices.ondevicechange? I think the following pattern might be useful:

1. A user plugs in a headset.
2. MediaDevices.ondevicechange refreshes the connected device.
3. The app will compare the new list with the previous list.
4. If a new device is connected (device.label changed), bubble up a dialog "is it a pair of headphones"?
5. The user makes a choice.

Of course this will interrupt the user's focus, but also it may be 99% of cases because of the "unknown" status anyway.

I'll summarize my findings soon.

**Comment by pmlt** [comment]
As discussed in the F2F meeting:

For the use case of deciding whether to enable 3D audio features, Wwise has moved away from headphone detection in favor of feature-detection of 3D audio support on the platform. 

Taking Windows as the primary example, instead of using IMMDevice API to retrieve the physical form property of the device (which is unreliable), instead we detect whether Microsoft Spatial Sound is active on the chosen output device (which is a user option in Windows). If Spatial Sound is enabled, we will detect the type of backend used (Windows Sonic vs. Dolby Atmos for example) and choose different rendering decisions based on that.

The point is that enabling Spatial Sound is an explicit action from the user, which is much more reliable.

On PS4, 3D audio can only be used if the user is using PS VR, so there is a programmatic way to detect that. Obviously the choice of using PS VR is also an explicit user action.

On PS5, the user can enable 3D audio in the console's dashboard, so we can detect that programmatically. Once again, this is an explicit user action from the user.

So for WebAudio we need to think about what is the corresponding "user action" to opt into 3D audio for the user. I could see just deferring this to an application-level option, or the platform-level choice is surfaced via WebAudio (which would be in the scope of issue #88).

**Comment by hoch** [comment]
I think this feature can still be useful for the streaming scenario. By detecting a personal listening device (e.g. headphones), you can save significant bandwidth. This makes more sense if the binauralization happens on the server-side.
